storagestate/storagestate.go

```diff
@@ -0,0 +1,730 @@
+type SetupParams struct {
+	aostypes.InstanceIdent
+	UID          int
+	GID          int
+	StateQuota   uint64
+	StorageQuota uint64
+}
+
+// StateChangedInfo contains state changed information.
+type StateChangedInfo struct {
+	aostypes.InstanceIdent
+	Checksum []byte
+}
+
+// StorageState storage state instance.
+type StorageState struct {
+	sync.Mutex
+	messageSender       MessageSender
+	storageDir          string
+	stateDir            string
+	storage             Storage
+	statesMap           map[aostypes.InstanceIdent]*stateParams
+	watcher             *fsnotify.Watcher
+	newStateChannel     chan cloudprotocol.NewState
+	stateRequestChannel chan cloudprotocol.StateRequest
+	isSamePartition     bool
+}
+
+type stateParams struct {
+	instancePath       string
+	stateFilePath      string
+	quota              uint64
+	checksum           []byte
+	changeTimer        *time.Timer
+	changeTimerChannel chan bool
+}
+
+/***********************************************************************************************************************
+ * Public
+ **********************************************************************************************************************/
+
+// New creates storagestate instance.
+func New(cfg *config.Config, messageSender MessageSender, storage Storage) (storageState *StorageState, err error) {
+	log.Debug("Create storagestate")
+
+	storageState = &StorageState{
+		storage:             storage,
+		storageDir:          cfg.StorageDir,
+		stateDir:            cfg.StateDir,
+		messageSender:       messageSender,
+		statesMap:           make(map[aostypes.InstanceIdent]*stateParams),
+		newStateChannel:     make(chan cloudprotocol.NewState, stateChannelSize),
+		stateRequestChannel: make(chan cloudprotocol.StateRequest, stateChannelSize),
+	}
+
+	if err = os.MkdirAll(storageState.storageDir, 0o755); err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	if err = os.MkdirAll(storageState.stateDir, 0o755); err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	if storageState.watcher, err = fsnotify.NewWatcher(); err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	storageMountPoint, err := fs.GetMountPoint(cfg.StorageDir)
+	if err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	stateMountPoint, err := fs.GetMountPoint(cfg.StateDir)
+	if err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	storageState.isSamePartition = storageMountPoint == stateMountPoint
+
+	go storageState.processWatcher()
+
+	return storageState, nil
+}
+
+// Close close storagestate instance.
+func (storageState *StorageState) Close() {
+	log.Debug("Close storagestate")
+
+	storageState.watcher.Close()
+}
+
+// Setup setup storagestate instance.
+func (storageState *StorageState) Setup(
+	params SetupParams,
+) (storagePath string, statePath string, err error) {
+	storageState.Lock()
+	defer storageState.Unlock()
+
+	instancePath := fmt.Sprintf("%d_%s_%s", params.Instance, params.SubjectID, params.ServiceID)
```

Comment:
it would be too long path. Probably better to generate some UUID? Or random dir and random file for storage and state but in this case. both values should be stored in db.

Reply:
Done. Use UUID generation